Gini-iOS-SDK/GINIURLSessionDelegate.m

```diff
@@ -9,42 +9,55 @@
++ (instancetype)urlSessionDelegateWithCertificatePaths:(NSArray<NSString *> *)certificatePaths {
+    return [[self alloc] initWithNSURLSessionDelegate:certificatePaths];
 }
 
-- (instancetype)initWithNSURLSessionDelegate:(NSString *)certificatePath {
+- (instancetype)initWithNSURLSessionDelegate:(NSArray<NSString *> *)certificatePaths {
     self = [super init];
     if (self) {
-        _nsCertificatePath = certificatePath;
+        _nsCertificatePaths = certificatePaths;
     }
     return self;
 }
 
 -(void)URLSession:(NSURLSession *)session
 didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge
 completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition, NSURLCredential * _Nullable))
-completionHandler {
-    
+completionHandler {    
     SecTrustRef serverTrust = challenge.protectionSpace.serverTrust;
-    SecCertificateRef certificate = SecTrustGetCertificateAtIndex(serverTrust, 0);
-    
-    NSMutableArray *policies = [NSMutableArray array];
-    [policies addObject:(__bridge_transfer id)SecPolicyCreateSSL(true, (__bridge CFStringRef)
-                                                                 challenge.protectionSpace.host)];
-    SecTrustSetPolicies(serverTrust, (__bridge CFArrayRef)policies);
+    BOOL remoteCertEqualToLocalCert = false;
+    BOOL remoteCertificateIsValid = false;
     
-    SecTrustResultType result;
-    SecTrustEvaluate(serverTrust, &result);
-    BOOL certificateIsValid = (result == kSecTrustResultUnspecified || result == kSecTrustResultProceed);
-    
-    NSData *remoteCertificateData = CFBridgingRelease(SecCertificateCopyData(certificate));
-    NSData *localCertificate = [NSData dataWithContentsOfFile:_nsCertificatePath];
+    for (int i = 0; i < SecTrustGetCertificateCount(serverTrust); i++) {
+        NSMutableArray *policies = [NSMutableArray array];
+        [policies addObject:(__bridge_transfer id)SecPolicyCreateSSL(true, (__bridge CFStringRef)
+                                                                     challenge.protectionSpace.host)];
+        SecTrustSetPolicies(serverTrust, (__bridge CFArrayRef)policies);
+        
+        SecTrustResultType result;
+        SecTrustEvaluate(serverTrust, &result);
+        remoteCertificateIsValid = (result == kSecTrustResultUnspecified || result == kSecTrustResultProceed);
+        
+        SecCertificateRef certificate = SecTrustGetCertificateAtIndex(serverTrust, i);
+        NSData *remoteCertificateData = (NSData *)CFBridgingRelease(SecCertificateCopyData(certificate));
+        
+        for (int j = 0; j < [_nsCertificatePaths count]; j++) {
+            NSData *localCertificate = [NSData dataWithContentsOfFile:[_nsCertificatePaths objectAtIndex:j]];
+            if(localCertificate != nil) {
+                remoteCertEqualToLocalCert = [remoteCertificateData isEqualToData:localCertificate];
+                if (remoteCertEqualToLocalCert) {
+                    goto outer_done;
```

Comment:
Goto??? Why?

Reply:
Why not? AFAIK you can use `break` with a boolean or just a labeled property like this. In my opinion it is more readable and clean than the other solution. If we had complexity O(n) instead of O(N2), I would have gone for the `break` approach.

Reply:
I have two problems with this:

Even if you want to make it "faster" by nit having a break, you have have a sub method that returns (because a return will break both loops).

But moreover, for these array sizes the complexity is not really a valid consideration. I would never sacrifice the code structure to spare two loop runs containing basic comparison. I would like to remind you that optimisation needs to be measured and things like "premature optimisations" definitely exist. Don't optimise unless you see the improvements.

That being said, I like that you are thinking about complexities and trying to write smarter code.